### PR TITLE
Cuda enabled - SimulatorCUDA

### PR DESCRIPTION
**Pull request by jaeyoo**
This PR enables tensor injection into SimulatorCUDA. (a tensor from TFQ context)

### CODE REVIEW

File: lib/simulator_cuda.h
Comment (sergeisakov) [inline lib/simulator_cuda.h:57]: Why do you need to inject this buffer?

Comment (jaeyoo) [reply lib/simulator_cuda.h:57]: That's old code legacy, let me remove it :) I had to inject it from a temporary buffer from Tensorflow Quantum context before, but proved that it's useless if we don't use TensorFlow GPU context, and then we don't have any conflict between qsim GPU context and tensorflow.

**Comment by jaeyoo** [comment]
macosx numpy versions : 1.21.0, 1.21.1, 1.21.2, 1.21.3, 1.21.4, 1.21.5, 1.21.6
manylinux numpy versions: 1.21.0, 1.21.1, 1.21.2, 1.21.3, 1.21.4, 1.21.5, 1.21.6
windows numpy versions : 1.21.0, 1.21.1, 1.21.2, 1.21.3, 1.21.4, 1.21.5, 1.21.6

Let me try 1.21.6.

**Comment by jaeyoo** [comment]
numpy failed at ubuntu/windows with 
1.21.0, 1.21.1, 1.22.0, 1.22.1, 1.22.2, 1.22.3, 1.22.4, 1.23.0rc1, 1.23.0rc2, 1.23.0rc3, 1.23.0, 1.23.1, 1.23.2, 1.23.3, 1.23.4, 1.23.5, 1.24.0rc1, 1.24.0rc2, 1.24.0, 1.24.1, 1.24.2

mac success with 1.21.6

let me try `1.21.1`

**Comment by jaeyoo** [comment]
compile error : `NumPy 1.21.1 may not yet support Python 3.11`

**Comment by jaeyoo** [comment]
roll back to the original numpy setup (`numpy~=1.16`)

File: .bazelrc
Comment (95-martin-orion) [inline .bazelrc:4]: From the [bazel docs](https://bazel.build/rules/lib/globals#parameters_36):

> Experimental. This parameter is experimental and may change at any time. Please do not depend on it.

File: WORKSPACE
Comment (95-martin-orion) [inline WORKSPACE:30]: [No change needed] I distinctly recall trying to use this stack of TF workspaces in the past, but it didn't work. Any idea what makes it possible now?

File: build_tools/test_libs.sh
Comment (95-martin-orion) [inline build_tools/test_libs.sh:22]: Why did these change?

**Review by 95-martin-orion — CHANGES_REQUESTED** [review]
Pinging @sergeisakov to confirm that `simmux.h` changes look OK.

File: .bazelrc
Comment (jaeyoo) [reply .bazelrc:4]: These are required in the TensorFlow bazel build at bazel 5.3.0 and TF v2.11.0

reference:

https://github.com/tensorflow/tensorflow/blob/master/.bazelrc#L438-L439

(it's defined common, which is applied to all options)

File: build_tools/test_libs.sh
Comment (jaeyoo) [reply build_tools/test_libs.sh:22]: because there is no `hybrid_test` and `qtrajectory_test`...?

File: WORKSPACE
Comment (jaeyoo) [reply WORKSPACE:30]: that was just TF-side error... it requires pre-installed dependency libraries (e.g. numpy, etc). so I added a line to install them like `python3 -m pip install -r requirements.txt` in the internal infra `kokoro_build.sh`.

File: build_tools/test_libs.sh
Comment (95-martin-orion) [reply build_tools/test_libs.sh:22]: Huh. Wonder how long it's been like that...

**Review by sergeisakov — CHANGES_REQUESTED** [review]

File: lib/simmux.h
Comment (jaeyoo) [reply lib/simmux.h]: Thank you for your comment, Sergei. I added `simmux_gpu.h`, and made `simulator_cuda` as default of gpu. so no more `__CUDA__` is required.

**Review by sergeisakov — COMMENTED** [review]
LGTM. Please wait for @95-martin-orion's approval.

**Review by sergeisakov — APPROVED** [review]

**Review by 95-martin-orion — APPROVED** [review]

**Pull request merged**
